doc/decisions/0_drafts/errors_core.md

```diff
@@ -0,0 +1,292 @@
+# Errors without `errorKey`
+
+## Problem
+
+We often use a `Key * errorKey` argument to return error information via metadata (see related decisions).
+In `libelektra-core`, however, that would be inappropriate and sometimes even impossible (e.g., `keyNew`).
+
+Not using a `Key * errorKey` argument can be limiting, however.
+Depending on the solution, there can be limited error values, i.e., it may not be possible to use a different error value for every possible error case.
+A prime example of this, a function that returns a pointer and wants to indicate an error via that pointer, only has to option to return `NULL`.
+So if there is more than one possible error, the error cases have to be distinguished another way.
+
+Therefore, we need a general concept for errors, which we use when an `errorKey` cannot be used.
+
+## Constraints
+
+- Using special return values as errors sometimes is limited.
+  For example, only `NULL` is an invalid pointer and sometimes even that is a legitimate non-error result.
+- It should be easy to detect error cases.
+- It must be possible to distinguish between errors, once an error is detected.
+- It should be hard to miss the fact that errors are possible.
+  Note, however, this does not mean checking for an error should be enforced.
+  There are many cases, where the caller knows an error is impossible, because of the previous code path.
+  The requirement is analogous to the `Key * errorKey` argument.
+  The presence of the argument makes it obvious that errors _can_ happen, but you can easily omit the error check when you want to.
+- Adding error indications should not make the API less convenient to use.
+
+## Assumptions
+
+- In many cases, it is possible to avoid error cases through the preceding code path.
+  For example, a very common case is that functions return an error, if they receive a `NULL` pointer.
+  Often this case can be excluded by static analysis, i.e., because of the preceding code it is impossible that the pointer is `NULL`.
+
+  ```c
+  int foo (KeySet * ks, Key * key) {
+    if (key == NULL) {
+      return -1;
+    }
+
+    // [...] could lots of code, as long as the Key * key doesn't change
+
+    const Key * found1 = ksLookupByName (ks, "/foo", 0);
+    if (found1 != NULL) {
+      // no need to worry about key == NULL, because we checked earlier
+      // also no need to worry about ks == NULL, because the ksLookupByName would have returned NULL
+      const Key * found2 = ksLookup (ks, key, 0);
+    }
+  }
+  ```
+
+- It is not always necessary to report separate error codes, for _every_ error case.
+  In many cases, the caller can distinguish after the fact, e.g., `keySetName` fails when the given name is invalid and when the given `Key` has a read-only name.
+  This can be distinguished after the fact, by checking whether the `Key` has a read-only name with `keyIsLocked`.
```

Comment:
I agree it can be distinguished but it still might be a problem. I already saw many errors where it was unclear which is the case because the application failed to distinguish these cases and debugging was needed to find out.

A failed KeySetName because of the name could be an input error but the read-only name would be an logic error (wrong understanding of when KeySetName can be called).

I don't think we should put this under the carpet with an assumption but should search for a solution with the best API usability.

Reply:
> I already saw many errors where it was unclear which is the case because the application failed to distinguish these cases and debugging was needed to find out.

That sounds like an application problem not an API problem.

> A failed KeySetName because of the name could be an input error but the read-only name would be an logic error (wrong understanding of when KeySetName can be called).

I see two ways this could happen (assuming there is an error check at all):
1. The application developer didn't read the docs of `keySetName` properly and assumed an error always means invalid keyname. As long as our error cases are documented properly, this is clearly an issue with the application, not with the API.
2. The application developer did do everything properly, but _after_ they wrote their application, we added a new error case to `keySetName` and reused the same error value. _That_ is our problem. But one could argue that this is an API breaking change. However, this issue also happens with any other solution, if we reuse error values for unrelated errors _after the fact_. 

There is also the obvious issue that can always happen in C: The application developer might not be checking for errors at all. That's a problem we cannot avoid. So we simply can't write an API that forces callers to check for errors.

> [We] should search for a solution with the best API usability.

That's exactly my point. We can't make a general statement about the API. To get the best API, it has to be decided on a case by case basis.

That's also my argument against the "Redesign API" and "Require caller to check preconditions" options. IMO we will just regret such strict rules at some point and we'll have to revisit the decision, because there is no good option left.

Reply:
> We can't make a general statement about the API. To get the best API, it has to be decided on a case by case basis.

But we also cannot do arbitrary changes. There must be concepts behind the API, that are fulfilled consistently to the point. E.g. the return value 1 means success, 0 no change, -1 error. To change something like this, a decision can be a good method, so it is a good idea to have such a decision. At the moment the decision is just to unfocused to be useful.

Reply:
> There must be concepts behind the API, that are fulfilled consistently to the point.

That's the whole point of this decision: Coming up with these concepts/rules.

I'm pretty sure we are on the same page, here and are just "talking past each other". I'll resolve the thread.